### PR TITLE
Added packing shaders for buffers with unaligned padding.

### PR DESCRIPTION
**Pull request by emilwestergren**
@ziquanyu take a look

### CODE REVIEW

**Comment by emilwestergren** [comment]
@ziquanyu can you test this?

**Comment by emilwestergren** [comment]
Updated.

File: source/draw/gpu/opengl/Map/OpenGLMapPack.ooc [hunk 1]
```diff
@@ -81,6 +84,42 @@ OpenGLMapPackUv: class extends OpenGLMap {
...
+			}
+			vec2 rg = texture(texture0, shiftedCoor).rg;
+			shiftedCoor = vec2(fragmentTextureCoordinate[1].x, fragmentTextureCoordinate[1].y);
+			if (shiftedCoor.x < 0.0) {
+				shiftedCoor = vec2(1.0 + shiftedCoor.x, shiftedCoor.y - rowUnit);
+			}
+			vec2 ba = texture(texture0, shiftedCoor).rg;
+			outColor = vec4(rg.x, rg.y, ba.x, ba.y);
+		}"
+}
```
Comment (davidhesselbom) [inline source/draw/gpu/opengl/Map/OpenGLMapPack.ooc]: Any reason we can't `slurp` these?

Comment (emilwestergren) [inline source/draw/gpu/opengl/Map/OpenGLMapPack.ooc]: No but I will make big PR later where I structure all shaders into files and `slurp` so I didn't do that here.

File: source/draw/gpu/opengl/OpenGLContext.ooc [hunk 2]
```diff
@@ -146,11 +149,19 @@ OpenGLContext: class extends GpuContext {
...
+		map: GpuMap
+		if (source instanceOf?(OpenGLMonochrome))
+			map = this _packMonochrome
+		else if (source instanceOf?(OpenGLUv)) {
+			channels = 2
+			if (padding == 0)
+				map = this _packUv
+			else {
+				map = this _packUvPadded
+				map add("paddingOffset", padding as Float / (target size width * 4) as Float)
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: 2nd `as Float` is unnecessary right?

Comment (davidhesselbom) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Correct.

File: source/draw/gpu/opengl/Map/OpenGLMapPack.ooc (hunk 1 above)
Comment (davidhesselbom) [inline source/draw/gpu/opengl/Map/OpenGLMapPack.ooc]: Alright.

**Comment by davidhesselbom** [comment]
Waiting for @ziquanyu to mark this as peer reviewed.

**Comment by emilwestergren** [comment]
Updated.

File: source/draw/gpu/opengl/OpenGLContext.ooc [hunk 3]
```diff
@@ -146,11 +149,19 @@ OpenGLContext: class extends GpuContext {
 		}
 	}
 	update: func { this _backend swapBuffers() }
-	packToRgba: func (source: GpuImage, target: GpuImage, viewport: IntBox2D) {
+	packToRgba: func (source: GpuImage, target: GpuImage, viewport: IntBox2D, padding := 0) {
 		channels := 1
```
Comment (davidhesselbom) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: There's no way to store the number of channels in the `GpuImage`, is there?

Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Should be possible, I think it was that way before but I don't remember why it was removed.

Comment (davidhesselbom) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Okay. Doesn't need fixing here, but it would make this prettier and could be useful elsewhere, too.

Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Yes, could even be a part of the `Image` class? Can you create issue for it?

Comment (davidhesselbom) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Alright.

Comment (davidhesselbom) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: #823

**Pull request merged**